blog/2020-04-07-ModernJavaConstraintsNoDRL.adoc

```diff
@@ -0,0 +1,153 @@
+That's a huge scalability gain:
+
+image::constraintStreamIncrementalCalculation.png[]
+
+== Indexing
+
+When joining multiple types, just like an SQL `JOIN` operator,
+ConstraintStreams applies hash lookups on indexes to scale better:
+
+image::constraintStreamJoinWithJoiners.png[]
+
+== IDE support
+
+Because ConstraintsStreams are written in the Java language,
+they piggy-back on very strong tooling support.
+
+Code highlighting, code completion and debugging just work:
+
+=== Code highlighting
+
+DRL code in IntelliJ IDEA Ultimate:
+
+image::codeHighlightingDRL.png[]
+
+Java code using ConstraintStreams in IntelliJ IDEA Ultimate, for the same constraints:
+
+image::codeHighlightingConstraintStreams.png[]
+
+=== Code completion
+
+Code completion for ConstraintStreams:
+
+image::codeCompletionConstraintStreams.png[]
+
+Of course, all API methods have Javadocs.
+
+=== Debugging
+
+Add a breakpoint in ConstraintStream's `filter()`:
+
+image::codeDebuggingConstraintStreams1.png[]
+
+To diagnose issues while debugging:
+
+image::codeDebuggingConstraintStreams2.png[]
+
+== Java syntax
+
+Constraints written in Java with ConstraintStreams follow the Java Language Specification (JLS), for good or bad.
+Similar logic applies when using ConstraintStreams from Kotlin or Scala.
+
+When migrating between DRL and ConstraintStreams, be aware of some differences between DRL and Java:
+
+* A DRL's `==` operator translates  to `equals()` in Java.
+* Besides getters, DRL also allows MVEL expressions that translate into getters in Java.
+
+For example, this DRL has `name` and `==`:
+
+[source,drl]
+----
+rule "Don't assign Ann"
+    when
+        Employee(name == "Ann")
+    then ...
+end
+----
+
+But the Java variant for the exact same constraint has `getName()` and `equals()` instead:
+
+[source,java]
+----
+constraintFactory.from(Employee.class)
+        .filter(employee -> employee.getName().equals("Ann"))
+        .penalize("Don't assign Ann", ...);
+----
+
+== Advanced functions
+
+The ConstraintStreams API allows us to add syntactic sugar
+and powerful new concepts, specifically tailored to help you build complex constraints.
+
+Just to highlight one of these, let's take a look at the powerful `groupBy` method:
+
+image::constraintStreamGroupBy.png[]
+
+Similar to an SQL `GROUP BY` operator or a Java 8 Stream `Collector`,
+it support `sum()`, `count()`, `countDistinct()`, `min()`, `max()`, `toList()`
+and even custom functions, again without loss of incremental score calculation.
+
+== Future work for ConstraintStreams
+
+First off, a big thanks to Lukáš Petrovický for all his work on ConstraintsStreams!
+
+But this is just the beginning.
+We envision more advanced functions,
+such as load balancing/fairness methods
+to make such constraints easier to implement.
+
+Right now, our first priority is to make it easier to unit test constraints in isolation.
+Think TDD (Test Driven Design). Stay tuned!
```

Comment:
Personally, I don't see the point of using the abbreviation, if all you do is then explain it in the same sentence. :-) Pick one.